[quant] adding memoryless observers for embeddingbag QAT work

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65443
* __->__ #65699

Summary: 
related to: https://github.com/pytorch/pytorch/pull/65443#discussion_r715132425

The QAT and PAT (pruning aware training) support for embedding bags needs a memoryless observer to work properly. This is necessitated by the changing pruned/non-pruned weights during training which can significantly change the quantization parameters.

This PR adds a memoryless flag to the simpler observer classes (not moving average since those explicitly have memory)

In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that it would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

python test/test_quantization.py TestObserver

(added test_memoryless_minmaxobserver, test_memoryless_per_channel_minmaxobserver, test_memoryless_histogramobserver)

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31209773](https://our.internmc.facebook.com/intern/diff/D31209773)

HDCharles committed:
[quant] adding memoryless observers for embeddingbag QAT work

Summary: Adding a memoryless flag to the non movingaverage observers so
they can be used similar to dynamic quant. In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that I would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] adding memoryless observers for embeddingbag QAT work"


Summary: 
related to: https://github.com/pytorch/pytorch/pull/65443#discussion_r715132425

Adding a memoryless flag to the observer classes so
they can be used similar to dynamic quant. This includes MovingAverageMinMaxObserver and MovingAveragePerChannelMinMaxObserver even though it is kind of a contradiction. This was done for consistency. In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that it would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

python test/test_quantization.py TestObserver

(added test_memoryless_minmaxobserver, test_memoryless_per_channel_minmaxobserver, test_memoryless_histogramobserver)

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] adding memoryless observers for embeddingbag QAT work"


Summary: 
related to: https://github.com/pytorch/pytorch/pull/65443#discussion_r715132425

Adding a memoryless flag to the observer classes so
they can be used similar to dynamic quant. This includes MovingAverageMinMaxObserver and MovingAveragePerChannelMinMaxObserver even though it is kind of a contradiction. This was done for consistency. In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that it would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

python test/test_quantization.py TestObserver

(added test_memoryless_minmaxobserver, test_memoryless_per_channel_minmaxobserver, test_memoryless_histogramobserver)

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/observer.py:
```diff
@@ -454,8 +458,8 @@ def extra_repr(self):
-        self.max_val = torch.tensor(float("-inf"))
+        self.min_val.copy_(torch.tensor(float("inf")))
+        self.max_val.copy_(torch.tensor(float("-inf")))
 
```
this is where the change to preserve the characteristics (device most importantly) of self.min_val and self.max_val was made since the way those values are assigned in the init statement uses factory_kwargs

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31209773).

HDCharles committed:
Update on "[quant] adding memoryless observers for embeddingbag QAT work"


Summary: 
related to: https://github.com/pytorch/pytorch/pull/65443#discussion_r715132425

Adding a memoryless flag to the observer classes so
they can be used similar to dynamic quant. This includes MovingAverageMinMaxObserver and MovingAveragePerChannelMinMaxObserver even though it is kind of a contradiction. This was done for consistency. In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that it would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

python test/test_quantization.py TestObserver

(added test_memoryless_minmaxobserver, test_memoryless_per_channel_minmaxobserver, test_memoryless_histogramobserver)

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31209773](https://our.internmc.facebook.com/intern/diff/D31209773)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31209773).

vkuzo commented on torch/ao/quantization/observer.py:
```diff
@@ -397,6 +397,7 @@ def __init__(
         dtype=torch.quint8,
         qscheme=torch.per_tensor_affine,
         reduce_range=False,
+        memoryless=False,
```
nit: a new keyword argument is safer to add to the end of the argument list, to avoid breaking user code

vkuzo reviewed (commented):
could we add some high level context on why this is needed to the PR summary

HDCharles committed:
Update on "[quant] adding memoryless observers for embeddingbag QAT work"


Summary: 
related to: https://github.com/pytorch/pytorch/pull/65443#discussion_r715132425

The QAT and PAT (pruning aware training) support for embedding bags needs a memoryless observer to work properly. This is necessitated by the changing pruned/non-pruned weights during training which can significantly change the quantization parameters.

This PR adds a memoryless flag to the observer classes so
they can be used similar to dynamic quant. For consistency, this includes MovingAverageMinMaxObserver and MovingAveragePerChannelMinMaxObserver even though a memoryless moving average observer is the same as a memoryless observer of its base type. 

In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that it would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

python test/test_quantization.py TestObserver

(added test_memoryless_minmaxobserver, test_memoryless_per_channel_minmaxobserver, test_memoryless_histogramobserver)

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31209773](https://our.internmc.facebook.com/intern/diff/D31209773)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31209773).

HDCharles committed:
Update on "[quant] adding memoryless observers for embeddingbag QAT work"


Summary: 
related to: https://github.com/pytorch/pytorch/pull/65443#discussion_r715132425

The QAT and PAT (pruning aware training) support for embedding bags needs a memoryless observer to work properly. This is necessitated by the changing pruned/non-pruned weights during training which can significantly change the quantization parameters.

This PR adds a memoryless flag to the observer classes so
they can be used similar to dynamic quant. For consistency, this includes MovingAverageMinMaxObserver and MovingAveragePerChannelMinMaxObserver even though a memoryless moving average observer is the same as a memoryless observer of its base type. 

In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that it would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

python test/test_quantization.py TestObserver

(added test_memoryless_minmaxobserver, test_memoryless_per_channel_minmaxobserver, test_memoryless_histogramobserver)

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31209773](https://our.internmc.facebook.com/intern/diff/D31209773)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31209773).

vkuzo commented:
> The QAT and PAT (pruning aware training) support for embedding bags needs a memoryless observer to work properly. This is necessitated by the changing pruned/non-pruned weights during training which can significantly change the quantization parameters.

thanks!  Could we also add context on which specific observers are needed for this use case?  If it's only `MinMaxObserver`, then it might be good to consider subclassing it instead of adding a setting to various classes.

HDCharles committed:
Update on "[quant] adding memoryless observers for embeddingbag QAT work"


Summary: 
related to: https://github.com/pytorch/pytorch/pull/65443#discussion_r715132425

The QAT and PAT (pruning aware training) support for embedding bags needs a memoryless observer to work properly. This is necessitated by the changing pruned/non-pruned weights during training which can significantly change the quantization parameters.

This PR adds a memoryless flag to the observer classes so
they can be used similar to dynamic quant. For consistency, this includes MovingAverageMinMaxObserver and MovingAveragePerChannelMinMaxObserver even though a memoryless moving average observer is the same as a memoryless observer of its base type. 

In addition to the above, I altered the reset_min_max_vals
function for MinMaxObserver so that it would preserve the device of the
existing self.min_val and self.max_val which was not preserved
previously compared to how it is initialized (using factory_kwargs)

Test Plan:

python test/test_quantization.py TestObserver

(added test_memoryless_minmaxobserver, test_memoryless_per_channel_minmaxobserver, test_memoryless_histogramobserver)

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31209773](https://our.internmc.facebook.com/intern/diff/D31209773)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31209773).

jerryzh168 commented:
is this for QAT dynamic quantization?

supriyar commented:
> is this for QAT dynamic quantization?

I think it would work for that, in which case we can probably connect this PR with T99683050

supriyar commented on torch/ao/quantization/observer.py:
```diff
@@ -400,6 +400,7 @@ def __init__(
         quant_min=None,
         quant_max=None,
         factory_kwargs=None,
+        memoryless=False,
```
can we add a docstring for `memoryless`

supriyar reviewed (approved):
lg. Can we update the description to remove MovingAvg observers?
